### PR TITLE
[Storage-Blob] Update type hinting `_blob_service_client`

### PR DESCRIPTION
**Pull request by neoseurae12**
# Description

update python type hinting in _blob_service_client.py

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md##building-and-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/storage/azure-storage-blob/azure/storage/blob/_blob_service_client.py [hunk 1]
```diff
@@ -619,17 +610,16 @@ def delete_container(
                 :dedent: 12
                 :caption: Deleting a container in the blob service.
         """
-        container = self.get_container_client(container) # type: ignore
```
Comment (vincenttran-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_blob_service_client.py:622]: Again, not sure if we can just remove it (or did your type checker mark these as unnecessary?)

File: sdk/storage/azure-storage-blob/azure/storage/blob/_blob_service_client.py [hunk 2]
```diff
@@ -619,17 +610,16 @@ def delete_container(
...
+        container = self.get_container_client(container)  
         kwargs.setdefault('merge_span', True)
         timeout = kwargs.pop('timeout', None)
-        container.delete_container( # type: ignore
```
Comment (vincenttran-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_blob_service_client.py:625]: Same comment as always about ignore

File: sdk/storage/azure-storage-blob/azure/storage/blob/_blob_service_client.py [hunk 3]
```diff
@@ -785,7 +772,7 @@ def get_blob_client(
             transport=TransportWrapper(self._pipeline._transport), # pylint: disable = protected-access
             policies=self._pipeline._impl_policies # pylint: disable = protected-access
         )
-        return BlobClient( # type: ignore
```
Comment (vincenttran-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_blob_service_client.py:788]: One more comment about ignore!

**Review by vincenttran-msft — COMMENTED** [review]
Hi @neoseurae12, thanks for your contribution! Overall the changes look good, just some minor comments on this first file.

In general I am still unsure what to do about the ` #type: ignore` so please inform me as to why you removed them (for example: the type checker you used said they were unnecessary or you just removed all of them 😄). 

The next steps for this pull-request would be:

1.  Please make all the necessary type hinting changes in `_blob_service_client_async.py` as well.
2. Getting the "Build Analyze" step from the testing pipeline to pass. Please let me know if you need help identifying the error output!

Thanks again for your contribution, hope that we can work together to get this merged!

**Comment by neoseurae12** [comment]
> Hi @neoseurae12, thanks for your contribution! Overall the changes look good, just some minor comments on this first file.
> 
> In general I am still unsure what to do about the ` #type: ignore` so please inform me as to why you removed them (for example: the type checker you used said they were unnecessary or you just removed all of them 😄).
> 
> The next steps for this pull-request would be:
> 
> 1. Please make all the necessary type hinting changes in `_blob_service_client_async.py` as well.
> 2. Getting the "Build Analyze" step from the testing pipeline to pass. Please let me know if you need help identifying the error output!
> 
> Thanks again for your contribution, hope that we can work together to get this merged!

With '1. Please make all the necessary type hinting changes in `_blob_service_client_async.py` as well', I made every necessary type hinting changes but for `# type: ignore` also in `_blob_service_client_async.py`.

and there was an import missing of `FilteredBlob` so i add that!

File: sdk/storage/azure-storage-blob/azure/storage/blob/_blob_service_client.py [hunk 4]
```diff
@@ -215,15 +214,14 @@ def get_user_delegation_key(self, key_start_time,  # type: datetime
         try:
             user_delegation_key = self._client.service.get_user_delegation_key(key_info=key_info,
                                                                                timeout=timeout,
-                                                                               **kwargs)  # type: ignore
```
[the remark this answers is not in the document]
Comment (neoseurae12) [reply sdk/storage/azure-storage-blob/azure/storage/blob/_blob_service_client.py:218]: i removed them which Pylance complained 'Unnecessary "# type: ignore" comment'

**Comment by neoseurae12** [comment]
> Hi @neoseurae12, thanks for your contribution! Overall the changes look good, just some minor comments on this first file.
> 
> In general I am still unsure what to do about the ` #type: ignore` so please inform me as to why you removed them (for example: the type checker you used said they were unnecessary or you just removed all of them 😄).
> 
> The next steps for this pull-request would be:
> 
> 1. Please make all the necessary type hinting changes in `_blob_service_client_async.py` as well.
> 2. Getting the "Build Analyze" step from the testing pipeline to pass. Please let me know if you need help identifying the error output!
> 
> Thanks again for your contribution, hope that we can work together to get this merged!

The reason I removed `# type: ignore`s was, actually, i removed them after i took some cue from your last PR (link: https://github.com/Azure/azure-sdk-for-python/pull/29045/commits/6b563610b030a391beb61eb73e16f15c20e2fdf1)

File: sdk/storage/azure-storage-blob/azure/storage/blob/_blob_service_client.py (hunk 1 above)
Comment (neoseurae12) [reply sdk/storage/azure-storage-blob/azure/storage/blob/_blob_service_client.py:622]: yup, my type checker told me those are unnecessary comments

**Comment by neoseurae12** [comment]
@vincenttran-msft some Build Analyze steps were failed 🥲 Could you please let me know what' next for this pull-request?

File: sdk/storage/azure-storage-blob/azure/storage/blob/_blob_service_client.py [hunk 5]
```diff
@@ -188,11 +188,10 @@ def from_connection_string(
         return cls(account_url, credential=credential, **kwargs)
 
     @distributed_trace
-    def get_user_delegation_key(self, key_start_time,  # type: datetime
-                                key_expiry_time,  # type: datetime
-                                **kwargs  # type: Any
-                                ):
-        # type: (...) -> UserDelegationKey
+    def get_user_delegation_key(self, key_start_time: datetime,
```
[the remark this answers is not in the document]
Comment (vincenttran-msft) [reply sdk/storage/azure-storage-blob/azure/storage/blob/_blob_service_client.py]: Your newest change is almost there, just need an additional indent on the parameter lines (192-194) and get the closing parenthesis (195) to line up with the method name (see `from_connection_string` in the same file for a good reference)

File: sdk/storage/azure-storage-blob/azure/storage/blob/aio/_blob_service_client_async.py [hunk 6]
```diff
@@ -131,11 +131,11 @@ def __init__(
...
-    async def get_user_delegation_key(self, key_start_time,  # type: datetime
-                                      key_expiry_time,  # type: datetime
-                                      **kwargs  # type: Any
-                                      ):
-        # type: (...) -> UserDelegationKey
+    async def get_user_delegation_key(
+        self, key_start_time: datetime,
+        key_expiry_time: datetime,
+        **kwargs: Any
+    ) -> UserDelegationKey:
```
Comment (vincenttran-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/aio/_blob_service_client_async.py:138]: Same feedback as async, almost there on formatting!

**Review by vincenttran-msft — COMMENTED** [review]
Overall, this batch of changes looks good! 

Now that you have the formatting correct on `get_user_delegation_key`, can you also fix the other multi-line functions in the files? Example: `__init__` indentation is not correct, `from_connection_string`, `set_service_properties`, `list_containers`, etc. 
Thanks! 😄 

I see that you removed the ignores because Pylance complained it is unnecessary. Please do that in `async` file as well then if they are marked as unnecessary!

Lastly, to see the results of the analyze, you can do the following:

1. Click "Details" next to Build Analyze step
2. Click "View more details on Azure Pipelines"
3. Under "Analyze" look for "Run PyLint"

To find the errors, I usually search for "Your code has been rated" and the output will be above.

For the current PR's state, here is the link to Azure Pipelines: https://dev.azure.com/azure-sdk/public/_build/results?buildId=3047198&view=logs&jobId=b70e5e73-bbb6-5567-0939-8415943fadb9&j=b70e5e73-bbb6-5567-0939-8415943fadb9&t=5da75d26-b661-5189-46a3-f9dd247b5d85

And here is the Pylint errors:
```
************* Module azure.storage.blob._blob_service_client
azure/storage/blob/_blob_service_client.py:192: [E0601(used-before-assignment), BlobServiceClient.get_user_delegation_key] Using variable 'datetime' before assignment
azure/storage/blob/_blob_service_client.py:195: [E0601(used-before-assignment), BlobServiceClient.get_user_delegation_key] Using variable 'UserDelegationKey' before assignment
azure/storage/blob/_blob_service_client.py:327: [E0601(used-before-assignment), BlobServiceClient.set_service_properties] Using variable 'BlobAnalyticsLogging' before assignment
azure/storage/blob/_blob_service_client.py:328: [E0601(used-before-assignment), BlobServiceClient.set_service_properties] Using variable 'Metrics' before assignment
azure/storage/blob/_blob_service_client.py:330: [E0601(used-before-assignment), BlobServiceClient.set_service_properties] Using variable 'CorsRule' before assignment
azure/storage/blob/_blob_service_client.py:332: [E0601(used-before-assignment), BlobServiceClient.set_service_properties] Using variable 'RetentionPolicy' before assignment
azure/storage/blob/_blob_service_client.py:333: [E0601(used-before-assignment), BlobServiceClient.set_service_properties] Using variable 'StaticWebsite' before assignment
azure/storage/blob/_blob_service_client.py:411: [E0601(used-before-assignment), BlobServiceClient.list_containers] Using variable 'ContainerProperties' before assignment
azure/storage/blob/_blob_service_client.py:475: [E0601(used-before-assignment), BlobServiceClient.find_blobs_by_tags] Using variable 'FilteredBlob' before assignment
azure/storage/blob/_blob_service_client.py:512: [E0601(used-before-assignment), BlobServiceClient.create_container] Using variable 'PublicAccess' before assignment
azure/storage/blob/_blob_service_client.py:563: [E0601(used-before-assignment), BlobServiceClient.delete_container] Using variable 'BlobLeaseClient' before assignment
azure/storage/blob/_blob_service_client.py:732: [E0601(used-before-assignment), BlobServiceClient.get_blob_client] Using variable 'BlobProperties' before assignment
************* Module azure.storage.blob.aio._blob_service_client_async
azure/storage/blob/aio/_blob_service_client_async.py:599: [C0301(line-too-long), ] Line too long (134/120)
azure/storage/blob/aio/_blob_service_client_async.py:135: [E0601(used-before-assignment), BlobServiceClient.get_user_delegation_key] Using variable 'datetime' before assignment
azure/storage/blob/aio/_blob_service_client_async.py:138: [E0601(used-before-assignment), BlobServiceClient.get_user_delegation_key] Using variable 'UserDelegationKey' before assignment
azure/storage/blob/aio/_blob_service_client_async.py:270: [E0601(used-before-assignment), BlobServiceClient.set_service_properties] Using variable 'BlobAnalyticsLogging' before assignment
azure/storage/blob/aio/_blob_service_client_async.py:271: [E0601(used-before-assignment), BlobServiceClient.set_service_properties] Using variable 'Metrics' before assignment
azure/storage/blob/aio/_blob_service_client_async.py:273: [E0601(used-before-assignment), BlobServiceClient.set_service_properties] Using variable 'CorsRule' before assignment
azure/storage/blob/aio/_blob_service_client_async.py:275: [E0601(used-before-assignment), BlobServiceClient.set_service_properties] Using variable 'RetentionPolicy' before assignment
azure/storage/blob/aio/_blob_service_client_async.py:276: [E0601(used-before-assignment), BlobServiceClient.set_service_properties] Using variable 'StaticWebsite' before assignment
azure/storage/blob/aio/_blob_service_client_async.py:454: [E0601(used-before-assignment), BlobServiceClient.create_container] Using variable 'PublicAccess' before assignment
azure/storage/blob/aio/_blob_service_client_async.py:505: [E0601(used-before-assignment), BlobServiceClient.delete_container] Using variable 'BlobLeaseClient' before assignment
azure/storage/blob/aio/_blob_service_client_async.py:671: [E0601(used-before-assignment), BlobServiceClient.get_blob_client] Using variable 'BlobProperties' before assignment
```

To fix the "used before assignment", for imports that are imported under `if TYPE_CHECKING:`, you need to wrap them in quotes ("") when referencing them in code. So for example, to resolve the first issue, line 192 shoud look like:
`self, key_start_time: datetime,` -> `self, key_start_time: "datetime",`

So that should help with those errors, and for the line too long just throw an ignore 
`  # pylint: disable=line-too-long`

Please let me know if you have any other questions!

**Pull request closed without merging**